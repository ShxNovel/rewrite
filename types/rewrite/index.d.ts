export * from "./rewrite";
export * from "./plot";
export * from "./tools";
export * from "./text-line";
export * from "./motion-line";
