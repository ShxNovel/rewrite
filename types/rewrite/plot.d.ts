import type { TextLine } from "./text-line";
import { type MotionLine } from "./motion-line";
export type RewriteInfo = {
    title?: string;
    icon?: string;
};
/**
 * Creates a plot information object;
 * It can be used by {@link RewritePlot.use}.
 */
export declare function rInfo(userArgs: Partial<RewriteInfo>): Partial<RewriteInfo>;
export type RewritePlotAttribute = {
    /** The type of plot, defaults to "Talk" */
    type: "Talk" | "Aside" | "Choice";
    label?: string;
} & RewriteInfo;
export type RewritePlotLine = {
    text: TextLine;
    motion: MotionLine;
};
export type RewritePlot = RewritePlotAttribute & RewritePlotLine & {
    /**
     * Initializes a new plot instance
     * and pushes the plot to rewrite stack
     */
    make(args: Partial<RewritePlotLine>): RewritePlot;
    /**
     * With side-effects. \
     * Apply user arguments to current plot configuration
     */
    use(args: Partial<RewritePlotAttribute>): RewritePlot;
};
export declare function rPlot(userArgs: Partial<RewritePlot>): {
    /** The type of plot, defaults to "Talk" */
    type: "Talk" | "Aside" | "Choice";
    label?: string;
} & RewriteInfo & RewritePlotLine & {
    /**
     * Initializes a new plot instance
     * and pushes the plot to rewrite stack
     */
    make(args: Partial<RewritePlotLine>): RewritePlot;
    /**
     * With side-effects. \
     * Apply user arguments to current plot configuration
     */
    use(args: Partial<RewritePlotAttribute>): RewritePlot;
} & Partial<RewritePlot>;
