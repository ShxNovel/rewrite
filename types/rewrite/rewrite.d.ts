import type { RewritePlot } from "./plot";
export declare function rewriteBegin(name: string, path: string): void;
export declare function rewritePush(plot: RewritePlot): void;
export declare function rewriteEnd(): void;
