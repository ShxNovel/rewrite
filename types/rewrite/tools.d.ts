export declare function makeArgs<T1, T2>(defaultArgs: T1, userArgs: T2): {} & T1 & T2;
export declare function mergeArgs<T1 extends object, T2 extends object>(defaultArgs: T1, userArgs: T2): T1 & T2;
