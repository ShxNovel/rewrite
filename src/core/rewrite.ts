import type { RewritePlot } from "./plot";

const files = [];

let nowFile = {
    name: "",
    path: "",
    content: [] as RewritePlot[],
};

function initFile(name: string = "", path: string = "") {
    nowFile = {
        name,
        path,
        content: [] as RewritePlot[],
    };
}

export function rewriteBegin(name: string, path: string) {
    initFile(name, path);
}

export function rewritePush(plot: RewritePlot) {
    nowFile.content.push(plot);
}

export function rewriteEnd() {
    files.push(nowFile);
    initFile();
}
