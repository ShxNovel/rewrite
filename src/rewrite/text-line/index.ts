export * from "./text-line";
export * from "./text-addon";
