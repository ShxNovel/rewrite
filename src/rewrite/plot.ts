import { makeArgs, mergeArgs } from "./tools";
import { rewritePush } from "./rewrite";

import type { TextLine } from "./text-line";
import { rTextLine } from "./text-line";
import { rMotionLine, type MotionLine } from "./motion-line";

export type RewriteInfo = {
    title?: string;
    icon?: string;
};

/**
 * Creates a plot information object;
 * It can be used by {@link RewritePlot.use}.
 */
export function rInfo(userArgs: Partial<RewriteInfo>): Partial<RewriteInfo> {
    return makeArgs({}, userArgs);
}

export type RewritePlotAttribute = {
    /** The type of plot, defaults to "Talk" */
    type: "Talk" | "Aside" | "Choice";

    label?: string;
} & RewriteInfo;

export type RewritePlotLine = {
    text: TextLine;
    motion: MotionLine;
};

export type RewritePlot = RewritePlotAttribute &
    RewritePlotLine & {
        /**
         * Initializes a new plot instance
         * and pushes the plot to rewrite stack
         */
        make(args: Partial<RewritePlotLine>): RewritePlot;

        /**
         * With side-effects. \
         * Apply user arguments to current plot configuration
         */
        use(args: Partial<RewritePlotAttribute>): RewritePlot;
    };

export function rPlot(userArgs: Partial<RewritePlot>) {
    const defaultPlot: RewritePlot = {
        type: "Talk",

        text: rTextLine(),
        motion: rMotionLine(),

        make(args: Partial<RewritePlotLine> = {}) {
            const result = mergeArgs(this, args);
            rewritePush(result);
            return result;
        },

        use(args) {
            return Object.assign(this, args);
        },
    };

    return mergeArgs(defaultPlot, userArgs);
}
