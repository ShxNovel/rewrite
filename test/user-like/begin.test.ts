import * as fs from "fs";
import * as path from "path";

import { assert, expect, test, vi } from "vitest";
import * as file from "../../src/core/rewrite.ts";

function getAllFiles(dirPath: string, arrayOfFiles: string[] = []) {
    const files = fs.readdirSync(dirPath);

    files.forEach((file) => {
        const fullPath = path.join(dirPath, file);
        if (fs.statSync(fullPath).isDirectory()) {
            arrayOfFiles = getAllFiles(fullPath, arrayOfFiles);
        } else {
            arrayOfFiles.push(fullPath);
        }
    });

    return arrayOfFiles;
}

const allFiles = getAllFiles("./test/user-like/files");
const filesMap = allFiles.map((file) => {
    const resolvedPath = path.resolve(file);
    return resolvedPath;
});

test("some", async () => {
    let i = 0;
    for (const one of filesMap) {
        file.rewriteBegin(path.basename(one), allFiles[i]);
        await import(one);
        file.rewriteEnd();
        i++;
    }

    const result = file.RewriteFiles;
    // expect(result).toMatchSnapshot();
});
