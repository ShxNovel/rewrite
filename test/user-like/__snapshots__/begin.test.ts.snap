// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`begin 1`] = `
[
  {
    "content": [
      {
        "begin": [Function],
        "icon": "name",
        "motion": {
          "motionLine": [],
          "r": [Function],
        },
        "text": {
          "pause": [Function],
          "text": [Function],
          "textLine": [
            {
              "args": {
                "word": "Hello",
              },
              "type": "Text",
            },
            {
              "args": {
                "ms": 1000,
              },
              "type": "Pause",
            },
            {
              "args": {
                "word": " World",
              },
              "type": "Text",
            },
          ],
        },
        "title": "title",
        "type": "Talk",
        "use": [Function],
        "useInfo": [Function],
        "useMotion": [Function],
        "useText": [Function],
      },
      {
        "begin": [Function],
        "icon": "name",
        "motion": {
          "motionLine": [],
          "r": [Function],
        },
        "text": {
          "pause": [Function],
          "text": [Function],
          "textLine": [
            {
              "args": {
                "word": "Hello",
              },
              "type": "Text",
            },
            {
              "args": {
                "ms": 1000,
              },
              "type": "Pause",
            },
            {
              "args": {
                "word": " World",
              },
              "type": "Text",
            },
          ],
        },
        "title": "title2",
        "type": "Talk",
        "use": [Function],
        "useInfo": [Function],
        "useMotion": [Function],
        "useText": [Function],
      },
    ],
    "name": "a.ts",
    "path": "test\\user-like\\files\\a.ts",
    "plotFlow": false,
  },
  {
    "content": [],
    "name": "b.ts",
    "path": "test\\user-like\\files\\b.ts",
    "plotFlow": true,
  },
]
`;
